WordPress/gutenberg
# Block API: Fix the block 'edit' property validation

## What?
PR fixes the block's `edit` property validation when bootstrapping the block types, failing for memoized components.

## Why?
The main requirement for the `edit` property is to be a valid React component for the block editor to render, which usually means having the type of a `function`, but can be a special `object` when wrapped in React APIs.

I noticed this while working on #57173. The HoCs like `withSelect` will return `pure` or memoized components.

## How?
Use the [`isValidElementType`](https://github.com/facebook/react/blob/main/packages/shared/isValidElementType.js#L38) from the official [`react-is`](https://www.npmjs.com/package/react-is) library to check the `edit` property's validity correctly.

P.S. If this check is too extensive, we can create our custom utility using primitives provided by the library.

## Testing Instructions
1. Apply the patch provided below.
2. Open the post editor.
3. It will `The "edit" property must be a valid function.` error on the trunk.
4. Repeat the same step on this branch.
5. There should be no errors in the console.

### Patch
```diff
diff --git packages/block-library/src/paragraph/edit.js packages/block-library/src/paragraph/edit.js
index ac766f69dd8..43e40b3bc0e 100644
--- packages/block-library/src/paragraph/edit.js
+++ packages/block-library/src/paragraph/edit.js
@@ -22,6 +22,7 @@ import {
 } from '@wordpress/block-editor';
 import { createBlock } from '@wordpress/blocks';
 import { formatLtr } from '@wordpress/icons';
+import { memo } from '@wordpress/element';
 
 /**
  * Internal dependencies
@@ -183,4 +184,4 @@ function ParagraphBlock( {
 	);
 }
 
-export default ParagraphBlock;
+export default memo( ParagraphBlock );
```

### Testing Instructions for Keyboard
Same
